src/toil/lib/misc.py

```diff
@@ -243,3 +254,53 @@ def atomic_copyobj(src_fh, dest_path, length=16384):
     with AtomicFileCreate(dest_path) as dest_path_tmp:
         with open(dest_path_tmp, 'wb') as dest_path_fh:
             shutil.copyfileobj(src_fh, dest_path_fh, length=length)
+
+
+class CalledProcessErrorStderr(subprocess.CalledProcessError):
+    """Version of CalledProcessError that include stderr in the error message if it is set"""
+
+    def __str__(self):
+        if (self.returncode and (self.returncode < 0)) or (self.stderr is None):
+            return str(super())
+        else:
+            err = self.stderr if isinstance(self.stderr, str) else self.stderr.decode("ascii", errors="replace")
+            return "Command '%s' exit status %d: %s" % (self.cmd, self.returncode, err)
+
+
+def popen_communicate(cmd, input=None, timeout=None, **kwargs):
+    """Start a process with subprocess.Popen, run the process and return
+    (stdout, stderr). Raise an exception if the process exits non-zero. Kwargs
+    are as passed to subprocess.Popen, with input and timeout being passed to
+    communicate.  The PY3.6 Popen encode and error arguments are allowed and
+    implemented.  Generally, Toil should used encoding='latin1'.
```

Comment:
Shouldn't we be using `utf8` throughout instead of `latin1`? As soon as someone runs Toil against a grid engine scheduler that titles its columns in Chinese or something, we're going to get in trouble with `latin1`.

Reply:
I am not convinced latin1 is right either ;-)

The problem, which I should have documented, is that we don't want to compound the
error reporting problem by having garbage on stderr cause a decoding exception with an invalid UTF8 character.  latin1 can handle all bytes without error.

Probably better would be to use utf-8 and set error to ignore or replace.

I have kind of decide that this is change is both a bit more code than it needs to be and doesn't go far enough.  I am thinking of making a wrapper around the new subprocess.run function that all of toil will use to call process. This way we  would get logging even of all programs executed plus provide the recommended encoding/error as defaults.

subprocess is somewhat limited in that you really want stderr to be strings even if stdin/stdout is binary ..

sigh, why can't ASCII be the universal language

Reply:
If we want to handle all possible bytes, it would make sense to be passing around bytes objects; we can repr() them for logging. But yeah, ignoring errors in the decoding would solve the same problem too. I just want to avoid a situation where we have to re-encode the library function's return value and decode it as utf-8 because some batch system prints out happy emoji for completed jobs or something.

Reply:
it now reads as utf-8,  however it also sets locale to C, due to Perl locale issues on one system.